### ISSUE TITLE
Amazon-ebs with file provisioner drop the head when copying directories.

### ISSUE THREAD

**Issue by ninoles** [mitchellh/packer#3719]
A file provisioner with a directory as source will remove the head directory (or directories if it ends with a '/') during the copy, when using amazon-ebs as builder.  It works as expected with the docker builder.
- Packer Version 0.10.1 (from packer.io)
- Host platform: Debian GNU/Linux (a bit like testing).
- https://gist.github.com/ninoles/965b671eda064ce39324ce7de52d5c43

The script (packer-bug.sh) in the gist creates a template and some subdirectory for testing.  It reproduce both the good behavior (with docker) and the bug (with amazon-ebs). I include the logs for both case (good and bad).

**Comment by rickard-von-essen** [comment]
I'm not sure the `ssh` communicatior is wrong, it might be the `docker` communicatiors upload that is wrong. 

I don't have access to a computer for the moment, can you double check that the ami you use don't have `noexec` option set on `/tmp` since it since it logs:
```
2016/07/14 12:09:20 packer: 2016/07/14 12:09:20 starting remote command: chmod +x /tmp/script_9487.sh; PACKER_BUILD_NAME='amazon-ebs' PACKER_BUILDER_TYPE='amazon-ebs' /tmp/script_9487.sh
2016/07/14 12:09:21 packer: 2016/07/14 12:09:21 remote command exited with '1': chmod +x /tmp/script_9487.sh; PACKER_BUILD_NAME='amazon-ebs' PACKER_BUILDER_TYPE='amazon-ebs' /tmp/script_9487.sh
```

**Comment by rickard-von-essen** [comment]
The docs for `file` provisioner clearly states that directory head will not be included when copying a directory. 

> If the source, however, is /foo/ (a trailing slash is present), and the destination is /tmp, then the contents of /foo will be uploaded into /tmp directly.

**Comment by ninoles** [comment]
Well, here, is not the directory head (which is bug/) that is missing, is the subdir head (dir).
If my interpretation is incorrect, than there is still a bug, since the docker behaviour is different.

**Comment by ninoles** [comment]
About your first comment, I think the output of the remote script and the output of packer is just mess up.  The exit 1 came from the run of the /tmp/script_9487.sh.

**Comment by ninoles** [comment]
OK, I upload a more complete test case, with a very discriminating issue:

With this directory structure:

- bug/dir1/file1.txt
- bug/dir1/subdir1/file11.txt
- bug/dir2/file2.txt
- bug/dir2/subdir2/file22.txt

And the provisioner like:

`{ "type": "file", "source": "bug/", "target": "/tmp/packer/" }` 

You end up with both dir1 and dir2 merged:

- /tmp/packer/file1.txt
- /tmp/packer/subdir1/file11.txt
- /tmp/packer/file2.txt
- /tmp/packer/subdir2/file22.txt

which is really unexpected.

So, although the behavior of the one without slash could be right this one is quite surprising.  Docker communicator behavior is more likely, and I'm pretty sure is very closed to what rsync does.

**Comment by rickard-von-essen** [comment]
Yes I agree, there is definitely a bug here. I'm just not sure where and what is the correct behavior. But currently I don't have access to a computer, I'll investigate further when I have one.

**Comment by ninoles** [comment]
I find the same bug in the virtualbox build.  I guess is because they share the same communicator ?

**Comment by rickard-von-essen** [comment]
Yes this is in the communicator. Docker has its own.

**Comment by pdutourgeerling** [comment]
I had the same problem. Fixed it by creating the directory on the EC2 instance before transferring data to it.

My `provisioners` in `build.json` looks like this:
```
"provisioners": [
  {
    "type": "shell",
    "inline": ["mkdir /tmp/directory"]
  },
  {
    "type": "file",
    "source": "../path/to/directory/",
    "destination": "/tmp/directory"
  },
  ...
]
```

**Comment by mwhooker** [comment]
I think this is similar to the winrm issue here #3562

The file provisioner is supposed to act like rsync, so just need to make sure the docker communicator uses the same behavior

**Comment by sanyer** [comment]
I bumped into same issue today. But in a bit different manner: only one of two folders is being copied into destination folder on instance, while both of them doesn't have trailing slash.

**Comment by sanyer** [comment]
Ok, for some reason using `sftp` also stopped working reliably. Now it only copies one of folders without trailing slashes.

So, I'm using https://github.com/mitchellh/packer/issues/3719#issuecomment-250501658 work-around.

@mwhooker is there a way to run packer in debugger? I'm using vscode.

**Comment by mwhooker** [comment]
@sanyer looks like you can use delve, which is pretty neat https://github.com/Microsoft/vscode-go/wiki/Debugging-Go-code-using-VS-Code

and make sure you've got `PACKER_LOG=1` set as an env var for more verbose logging.

**Comment by mwhooker** [comment]
tl;dr: the fix is to make your first shell provisioner do:

```json
{
    "type": "shell",
    "inline": [
        "mkdir -p /tmp/packer/without-slash",
        "mkdir -p /tmp/packer/with-slash"
    ]
}
```

There appears to be a bug with your script. The destination path _must_ exist
for the file upload to work correctly. Since the directories
/tmp/packer/{with,without}-slash don't exist, the behavior is essentially
undefined, but

With your script, the local contents were:

```
bug
bug/dir1
bug/dir1/file1.txt
bug/dir1/subdir1
bug/dir1/subdir1/file11.txt
bug/dir2
bug/dir2/file2.txt
bug/dir2/subdir2
bug/dir2/subdir2/file22.txt
```

and the uploaded contents were:

```
1489107421,,ui,message,    amazon-ebs: /tmp/packer
1489107421,,ui,message,    amazon-ebs: /tmp/packer/without-slash
1489107421,,ui,message,    amazon-ebs: /tmp/packer/without-slash/subdir1
1489107421,,ui,message,    amazon-ebs: /tmp/packer/without-slash/subdir1/file11.txt
1489107421,,ui,message,    amazon-ebs: /tmp/packer/without-slash/file1.txt
1489107421,,ui,message,    amazon-ebs: /tmp/packer/with-slash
1489107421,,ui,message,    amazon-ebs: /tmp/packer/with-slash/file2.txt
1489107421,,ui,message,    amazon-ebs: /tmp/packer/with-slash/subdir2
1489107421,,ui,message,    amazon-ebs: /tmp/packer/with-slash/subdir2/file22.txt
1489107421,,ui,message,    amazon-ebs: /tmp/packer/with-slash/subdir1
1489107421,,ui,message,    amazon-ebs: /tmp/packer/with-slash/subdir1/file11.txt
1489107421,,ui,message,    amazon-ebs: /tmp/packer/with-slash/file1.txt
```

Let's start with `without-slash`:

```
{
    "type": "file",
    "source": "{{template_dir}}/bug/dir1",
    "destination": "/tmp/packer/without-slash/"
}
```

Since the directory `/tmp/packer/without-slash` doesn't exist, the command is
interpreted as uploading the contents of the `dir1` directory to
`/tmp/packer/without-slash`, which is how `cp` and `scp` both work.

```
$ mkdir foo
$ scp -vr ./bug/dir1 ./foo/without-slash/
Executing: cp -r -- ./bug/dir1 ./foo/without-slash/
$ find foo
foo
foo/without-slash
foo/without-slash/file1.txt
foo/without-slash/subdir1
foo/without-slash/subdir1/file11.txt
```

I agree that this isn't documented, and will see about adding a note about it.

As for the `with-slash` directory, the behavior is a little harder to explain:

```
{
    "type": "file",
    "source": "{{template_dir}}/bug/",
    "destination": "/tmp/packer/with-slash/"
}
```

I believe that since there's a trailing slash, it descends into the directory
and uploads each subdirectory to `with-slash`, which means it overwrites the
destination directory each time.


I'm not quite clear on what the right solution is, here, since we can't easily
check that the remote destination exists, given differences in OSes. I think
the best we can do for now is make the docs extra clear. If anyone has any
ideas I'd be happy to hear them.

**Comment by mwhooker** [comment]
btw, thank you for the work of making this easily reproducible. much appreciated!
